torch/csrc/jit/script/compiler.cpp

```diff
@@ -1086,85 +1086,47 @@ struct to_ir {
-          lc.range(),
-          /*n_binders*/ 1);
-      return emitExpr(lc.elt())->type();
-    });
-  }
-
-  // emit a single index of the Iterable to get the children type
-  TypePtr getIterableChildrenType(const ListComp& lc, IterableTree& tree) {
-    return emitInTempEnvironment([&]() {
-      return tree.getitem(lc.range(), method, graph->insertConstant(0))
-          ->asValue(lc.range(), method)
-          ->type();
-    });
-  }
-
-  Value* emitListComprehension(const ListComp& lc) {
-    const auto tmp_name = createTempName("$list_acc");
-    const auto sv = emitSugaredExpr(lc.iter(), 1);
-
-    TypePtr list_type;
-
-    auto siv = dynamic_cast<SimpleValue*>(sv.get());
-    if (siv && siv->getValue()->type()->cast<ListType>()) {
-      auto list_elem =
-          siv->getValue()->type()->cast<ListType>()->getElementType();
-      list_type = getListCompType(lc, list_elem);
-    } else if (auto iterable_tree = dynamic_cast<IterableTree*>(sv.get())) {
-      list_type =
-          getListCompType(lc, getIterableChildrenType(lc, *iterable_tree));
-    } else if (dynamic_cast<RangeValue*>(sv.get())) {
-      list_type = getListCompType(lc, IntType::get());
-    } else {
-      throw ErrorReport(lc.range())
-          << "iterator expression is expected to be a list, iterable, or "
-             "range, found "
-          << sv->kind();
-    }
-
-    // given `[x*2 for x in my_list]` this generates the following AST:
-    // __list_acc = []
-    // for x in my_list:
-    //  __list_acc.append(x*2)
-    const auto n =
-        graph->insertNode(graph->createList(list_type, at::ArrayRef<Value*>{}));
-    environment_stack->setVar(lc.range(), tmp_name, n->output());
-    const auto tmp_list_ident = Ident::create(lc.range(), tmp_name);
-    const auto tmp_list_var = Var::create(lc.range(), tmp_list_ident);
-    const auto append_ident = Ident::create(lc.range(), "append");
-    const auto dot_op = Select::create(lc.range(), tmp_list_var, append_ident);
-    const auto append_args_list = List<Expr>::create(lc.range(), {lc.elt()});
-    const auto append_attrs = List<Attribute>::create(lc.range(), {});
-    const auto apply_append =
-        Apply::create(lc.range(), dot_op, append_args_list, append_attrs);
-    const auto expr_stmt = ExprStmt::create(lc.range(), apply_append);
-    const auto stmt_list = List<Stmt>::create(lc.range(), {expr_stmt});
-    const auto iters_list = List<Expr>::create(lc.range(), {lc.iter()});
+  Value* emitListComprehension(const ListComp& lc, const TypePtr& type_hint) {
+    const auto loc = lc.range();
     const auto targets_list = List<Expr>::create(lc.range(), {lc.target()});
-    const auto for_loop =
-        For::create(lc.range(), targets_list, iters_list, stmt_list);
-    emitFor(for_loop);
-    return n->output();
+    const auto itrs = List<Expr>::create(lc.range(), {lc.iter()});
+    auto placeholder_node = graph->insertNode(create(prim::Print, loc, 0));
```

Comment:
It is more clear if we just create a List node immediately and then change its type later.  The logic with a fake prim::Print node is weird and takes a long time to read to figure out what is going on.